File: hahlw/array/56.合并区间.cpp
/*
 * @lc app=leetcode.cn id=56 lang=cpp
 *
 * [56] 合并区间
 */
class Solution {
public:
    static bool cmp(vector<int>&a,vector<int>&b){
        if(a[0]<b[0]||(a[0]==b[0]&&a[1]<b[1]))return true;
        else return false;
    }
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(),intervals.end(),cmp);
        for (int i=0;i<intervals.size();i++){
             int start = intervals[i][0];
             int end = intervals[i][1];
             int j = i + 1;
             while(j<intervals.size()){
                if(end>=intervals[j][0]){//a[1]==b[0]
                    end=max(intervals[j][1],end);//a[0]<b[0]&&a[1]>b[1]
                    j++;
                }
                else break;
             }
             if(j!=i+1){
                 intervals.erase(intervals.begin()+i,intervals.begin()+j);
                 intervals.insert(intervals.begin()+i,{start,end});
             }
        }
        return intervals;
    }
};

File: Juniper/56.合并区间.py
#
# @lc app=leetcode.cn id=56 lang=python3
#
# [56] 合并区间
#
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        lens = len(intervals)
        i = 0
        intervals.sort(key=lambda x:x[0])
        while i < lens:
            j = i
            r = intervals[i][1]
            if i < lens-1 and r >= intervals[i+1][0]:
                while i < lens-1 and r >= intervals[i+1][0]:
                    if r < intervals[i+1][1]:
                        r = intervals[i+1][1]
                    i +=1
                res.append([intervals[j][0], r])
            else:
                res.append(intervals[i])
            i +=1    
        return res


        

